### PR TITLE
Tests/scu

### PR DESCRIPTION
**Pull request by boygirl**
cc/ @ryan-roemer 
addresses comments in https://github.com/FormidableLabs/victory-core/pull/272

### CODE REVIEW

File: test/client/spec/victory-util/collection.spec.js [hunk 1]
```diff
@@ -174,4 +177,71 @@ describe("collections", () => {
...
+    });
+
+    it("recursively checks equality for nested collections", () => {
+      const a = { test: { nested: "a" }, test2: { nested: "a" } };
+      const b = { test: { nested: "a" }, test2: { nested: "a" } };
+      expect(Collection.areVictoryPropsEqual(a, b)).to.equal(true);
+      expect(Collection.checkEquality).calledThrice;
+    });
+
+    it("does not recursively check date objects", () => {
```
Comment (ryan-roemer) [inline test/client/spec/victory-util/collection.spec.js]: This doesn't seem to be the right test title. These dates _are_ being compared, and there isn't a nested object. As luck has it, they're all equal, but aren't necessarily.

We probably need to (1) retitle and/or refactor this test, and (2) make sure to use fake timers in our sandbox, amending the`sandbox.create()` call above to:

```js
      sandbox = sinon.sandbox.create({
        useFakeTimers: true
      });
```

This will have all dates use unix epoch as "now" unless `sandbox.clock.tick(MS_AMOUNT)` is called to advance fake time.

File: test/client/spec/victory-util/collection.spec.js [hunk 2]
```diff
@@ -174,4 +177,71 @@ describe("collections", () => {
...
+      expect(Collection.checkEquality).calledOnce.and.returned(true);
+    });
+
+    it("returns equal for equivalent date objects", () => {
+      const a = { test: new Date(2010, 1, 1) };
+      const b = { test: new Date("Mon Feb 01 2010 00:00:00 GMT-0800 (PST)") };
+      expect(Collection.areVictoryPropsEqual(a, b)).to.equal(true);
+    });
+
+  });
```
Comment (ryan-roemer) [inline test/client/spec/victory-util/collection.spec.js]: Looking good generally, but I think we've got some corner cases to cover still. For example, we return `true` for `{}` vs. `[]` when that probably should be `false`. Here's a diff with that test and a whole bunch of lodash "isEqual" unit tests (_note_ we should not just copy them like in my diff, but use them as a starting point for our tests).

I think we'll want to look at this diff, and additional lodash isEqual unit tests for additional cases we should cover: https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9477

```diff
diff --git a/test/client/spec/victory-util/collection.spec.js b/test/client/spec/victory-util/collection.spec.js
index eeab6d4..4fa14d2 100644
--- a/test/client/spec/victory-util/collection.spec.js
+++ b/test/client/spec/victory-util/collection.spec.js
@@ -179,10 +179,12 @@ describe("collections", () => {
   });
 
 
-  describe("areVictoryPropsEqual", () => {
+  describe.only("areVictoryPropsEqual", () => {
     let sandbox;
     beforeEach(() => {
-      sandbox = sinon.sandbox.create();
+      sandbox = sinon.sandbox.create({
+        useFakeTimers: true
+      });
       sandbox.spy(Collection, "checkEquality");
     });
     afterEach(() => {
@@ -243,5 +245,39 @@ describe("collections", () => {
       expect(Collection.areVictoryPropsEqual(a, b)).to.equal(true);
     });
 
+    // FAILS
+    it("returns false for empty object vs. array", () => {
+      const a = { a: [] };
+      const b = { a: {} };
+      expect(Collection.areVictoryPropsEqual(a, b)).to.equal(false);
+      expect(Collection.checkEquality).calledOnce.and.returned(false);
+    });
+
+    // Disable symbol tests.
+    // https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9480-L9496
+    const symbol1 = true;
+    const symbol2 = false;
+    const pairs = [
+      [1, 1, true], [1, Object(1), true], [1, '1', false], [1, 2, false],
+      [-0, -0, true], [0, 0, true], [0, Object(0), true], [Object(0), Object(0), true], [-0, 0, true], [0, '0', false], [0, null, false],
+      [NaN, NaN, true], [NaN, Object(NaN), true], [Object(NaN), Object(NaN), true], [NaN, 'a', false], [NaN, Infinity, false],
+      ['a', 'a', true], ['a', Object('a'), true], [Object('a'), Object('a'), true], ['a', 'b', false], ['a', ['a'], false],
+      [true, true, true], [true, Object(true), true], [Object(true), Object(true), true], [true, 1, false], [true, 'a', false],
+      [false, false, true], [false, Object(false), true], [Object(false), Object(false), true], [false, 0, false], [false, '', false],
+      [symbol1, symbol1, true], [symbol1, Object(symbol1), true], [Object(symbol1), Object(symbol1), true], [symbol1, symbol2, false],
+      [null, null, true], [null, undefined, false], [null, {}, false], [null, '', false],
+      [undefined, undefined, true], [undefined, null, false], [undefined, '', false]
+    ];
+
+    // SOME FAIL
+    pairs.forEach((vals) => {
+      it(`lodash primitive comparison: ${vals}`, () => {
+        const a = vals[0];
+        const b = vals[1];
+        const expected = vals[2];
+        expect(Collection.areVictoryPropsEqual(a, b)).to.equal(expected);
+      });
+    });
+
   });
 });
```

**Review by ryan-roemer — CHANGES_REQUESTED** [review]
Thanks for getting this started so quickly!

Code changes look good. Tests cases I think could use more coverage, and I think we can use lodash's for inspiration!

File: test/client/spec/victory-util/collection.spec.js (hunk 1 above)
Comment (boygirl) [reply test/client/spec/victory-util/collection.spec.js]: You're right, these should be exact dates instead. I included this test to distinguish date objects from plain objects, as plain objects should be _recursively_ checked, and dates should not.

**Comment by boygirl** [comment]
This equality checker differs from lodash `isEqual` in the following ways:
- Functions are marked as equal (By design-- this was the main impetus for writing a new function)
- Does not handle symbols or maps
- Returns false when checking the equality of things like `1` vs. `Object(1)` whereas isEqual returns true
- Does not handle circular references in objects and arrays

File: test/client/spec/victory-util/collection.spec.js [hunk 3]
```diff
@@ -174,4 +177,104 @@ describe("collections", () => {
...
+      expect(Collection.checkEquality).calledOnce.and.returned(true);
+    });
+
+    it("returns equal for equivalent date objects", () => {
+      const a = { test: new Date(2010, 1, 1) };
+      const b = { test: new Date("Mon Feb 01 2010") };
+      expect(Collection.areVictoryPropsEqual(a, b)).to.equal(true);
+    });
+
+    const pairs = [
```
Comment (ryan-roemer) [inline test/client/spec/victory-util/collection.spec.js]: We need to either (1) attribute the lodash license and make sure it's compatible with Victory's for this literal copy and paste, OR (2) use it for inspiration / cases but not copy and paste it. Either way is fine.

File: test/client/spec/victory-util/collection.spec.js [hunk 4]
```diff
@@ -174,4 +177,104 @@ describe("collections", () => {
...
+
+    pairs.forEach((vals) => {
+      it(`matches lodash primitive comparison: ${vals}`, () => {
+        const a = vals[0];
+        const b = vals[1];
+        const expected = vals[2];
+        expect(Collection.areVictoryPropsEqual(a, b)).to.equal(expected);
+      });
+    });
+  });
```
Comment (ryan-roemer) [inline test/client/spec/victory-util/collection.spec.js:357]: These were only the "primitives" tests for lodashs isEqual. There are many others we should incorporate as well if they apply to scenarios we need to support.

File: src/victory-util/collection.js [hunk 5]
```diff
@@ -95,5 +95,50 @@ export default {
...
+
+  basicEqualityCheck(a, b) {
+    if (a === b) { return true; }
+    if (typeof a !== typeof b) { return false; }
+    // isEqual does not support equality checking on functions
+    // return true if a and b are both functions
+    if (typeof a === "function") { return true; }
+    if (typeof a === "object" && keys(a).length !== keys(b).length) { return false; }
+    if (typeof a !== "object" || keys(a).length === 0) { return isEqual(a, b); }
+    return undefined;
```
Comment (ryan-roemer) [inline src/victory-util/collection.js]: We should document this internal function to say that it's tri-state and spell out that you cannot just use it straight up without comparing the type of the return value.

Comment (boygirl) [reply src/victory-util/collection.js]: better to also make it an actual internal function?

Comment (ryan-roemer) [reply src/victory-util/collection.js]: Might be nice -- could add a few specific tests if warranted too.

**Comment by boygirl** [comment]
@ryan-roemer this is feeling complete enough to me. Good to merge?

**Comment by ryan-roemer** [comment]
@boygirl -- Did you see my question about https://github.com/FormidableLabs/victory-core/pull/273#discussion_r129379054 ?

Given that the lodash primitives examples yielded some good bugs should we look at their _other_ cases too? (Not just the ones I pasted for you, actually looking at their other isEqual tests like:

- https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9509
- https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9587
- https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9625
- https://github.com/lodash/lodash/blob/4.17.4/test/test.js#L9833

... and everything else that they're already done the work for...)

**Comment by boygirl** [comment]
Yep, [this commit](https://github.com/FormidableLabs/victory-core/pull/273/commits/abe53957c6cbad7a3303daab06907cc47956e4eb) added specs for equality checking for arrays, plain objects, nested arrays, nested objects, and mixed collections. I feel like those are the relevant checks given that this method will only ever be used for checking equality on validated props objects and a state object.

**Review by ryan-roemer — APPROVED** [review]

**Comment by ryan-roemer** [comment]
OK!
